### DISCUSSION TITLE
Google Seach Console integration issue

### DISCUSSION THREAD

**Discussion by wheeleran** [plausible/analytics#2926] [Self-Hosted Support]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Using official Plausible Cloud hosting or self-hosting?

Self-hosting

### Describe the bug

The Plausible docs do not mention that the Search Console integration set out in the docs is only valid for 7 days, after which the Google token expires and the Google account has to be unlinked and relinked. This qualification should be added to the docs because there is NO mention of it, or the docs updated to (hopefully) explain how to make a suitable live production app, rather than the simple testing app with the 7 day limitation which it currently describes. 

### Expected behavior

Change docs to explain that the described method is only valid for 7 days, or change docs to describe how to make a production app.

**Comment by Tony-MF** [comment]
This is not the case for me. I have made the connection once and it works fine 2 months later.

**Reply by wheeleran** [reply]
@Tony-MF this gives me hope; did you make a live or a testing app, because i've made a testing app and that apparently does have the 7 day token renewal issue i'm seeing. and can you talk me through the last bit, enabling the Google Search Console API and the Domain verification section? i've checked what i've done so far, and I just couldn't find "Domain verification" in the menu, and that's the thing i think i'm missing. But I've checked in Google Search Console API > manage, then clicked the name of my OAuth client, and the Authorized Redirect URI is there, with the '/auth/google/callback' ending. Is that it, or is there another Domain verification bit i'm missing? i followed the instructions [here](https://plausible.io/docs/self-hosting-configuration#enable-google-search-console-api), but I didn't get any prompts to go to the Google Search Console.

**Reply by jonahgreenthal** [reply]
I also can't find "Domain Verification" and would really appreciate help!

**Reply by bartse** [reply]
@jonahgreenthal Did you manage to find the solution for the missing `Domain Verification` ? Im missing that button in the Api & Services menu and cant figure out why.

**Reply by wheeleran** [reply]
Nope, and no feedback either, as you can see. Either I'm blind or Plausible's docs are misleading.

**Reply by raabmar** [reply]
Any news on this? I am also searching the `Domain verification` button and just can't finde how to make it work

**Comment by wheeleran** [comment]
> This is not the case for me. I have made the connection once and it works fine 2 months later. 

Could you explain if you found and where is the mysterious and elusive Domain Verification step mentioned in the Plausible docs?

**Comment by jonahgreenthal** [comment]
> @jonahgreenthal Did you manage to find the solution for the missing `Domain Verification` ? Im missing that button in the Api & Services menu and cant figure out why.

I don't remember every detail, but I did eventually get it worked out. I think the domain verification thing was just plain unnecessary, possibly related to this:
![image](https://github.com/plausible/analytics/assets/737172/1409daf3-d778-487f-90c2-2e945ae283b1)

**Reply by wheeleran** [reply]
But here's the thing: there's two whole points in the Plausible docs about "verifying the domain" and the instructions on where to find it and what to do are not findable. Apparently not just by me. There was no verifying the domain so for now it works, for seven days, then the keys expire as far as I can tell. And I made a production app too, not a test app. I'm sure it's me, but I'm hoping to be talked through it step by step.

**Reply by Normanras** [reply]
For some reason, I can't even get mine to work for the 7 days. I can see that my google app/api has given access to Plausible successfully. I've done this more than a few times before for HomeAssistant, so I'm familiar with Google Cloud console and Oauth consent, etc. However, I don't see any data flowing into Plausible despite the fact that it says "you've successfully connected!" to Google in Plausible's UI. 

In troubleshooting, there's also [these troubleshooting steps](https://plausible.io/docs/google-analytics-import#my-import-has-failed) . Specifically, the step to "Enable Users Metrics in Reporting" can't be found in Google. Google itself has a [support article](https://support.google.com/analytics/answer/2992042?hl=en#zippy=) with the same instructions, but maybe with the update to GA5 they deprecated the setting in most properties. 

So now I keep getting success messages, but no new data is showing... I might wipe my app in Google and start from the beginning...

**Comment by 0PandaDEV** [comment]
I have set up everything exactly how it is in the plausible docs but every time I want to connect a website in the dashboard to my Google account it gives me this error "Error 400: redirect_uri_mismatch"
![image](https://github.com/plausible/analytics/assets/70103896/81f153b4-2a72-4ac7-b055-33f964f80ec5)

**Reply by anpin** [reply]
have you found a solution for this?

**Reply by anpin** [reply]
apparently I had to use `http://` scheme for the callback url. However after successful authentication I get 
![image](https://github.com/plausible/analytics/assets/6060545/72afd26f-41f4-44df-80a8-c39426b280bd)

**Reply by ruslandoga** [reply]
👋 @anpin 

That's a bit too strange to be the fix. The issue is most likely in something else. Would you be able to share more info about your setup? Have you seen https://github.com/plausible/analytics/discussions/2926#discussioncomment-8568486?

**Reply by 0PandaDEV** [reply]
> have you found a solution for this?

Yes, just follow the new documentation of the [plausible community edition](https://github.com/plausible/community-edition/) and make sure you update to the latest version.

| From | To |
|------|------|
| [2.0.0](https://github.com/plausible/analytics/discussions/3132) | [2.1.0](https://github.com/plausible/analytics/discussions/4125) |
| [2.1.0](https://github.com/plausible/analytics/discussions/4125) | [2.1.1](https://github.com/plausible/analytics/discussions/4195) |

**Reply by anpin** [reply]
I'm on nixos and it is still on v2.0.0, will try again after it gets updated https://github.com/NixOS/nixpkgs/issues/314768

**Reply by anpin** [reply]
here are the logs 

```console
Jul 10 17:07:00 plausible plausible-start[982]: Request: GET /auth/google/callback?state=%5B1,%22search-console%22%5D&code=<token>&scope=email%20https://www.googleapis.com/auth/webmasters.readonly%20openid%20https://www.googleapis.com/auth/userinfo.email&authuser=0&prompt=consent
Jul 10 17:07:00 plausible plausible-start[982]: ** (exit) an exception was raised:
Jul 10 17:07:00 plausible plausible-start[982]:     ** (MatchError) no match of right hand side value: {:error, %Mint.TransportError{reason: :nxdomain}}
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible/google/http.ex:128: Plausible.Google.HTTP.fetch_access_token/1
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:598: PlausibleWeb.AuthController.google_auth_callback/2
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.action/2
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/controllers/auth_controller.ex:1: PlausibleWeb.AuthController.phoenix_controller_pipeline/2
Jul 10 17:07:00 plausible plausible-start[982]:         (phoenix 1.7.2) lib/phoenix/router.ex:430: Phoenix.Router.__call__/5
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.plug_builder_call/2
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint."call (overridable 3)"/2
Jul 10 17:07:00 plausible plausible-start[982]:         (plausible 0.0.1) lib/plausible_web/endpoint.ex:1: PlausibleWeb.Endpoint.call/2
```

**Reply by ruslandoga** [reply]
@anpin are you running on a IPv6-only network perhaps? Ah, no. `dig aaaa www.googleapis.com` works.

**Reply by ruslandoga** [reply]
The request fails here: https://github.com/plausible/analytics/blob/0621fc0a77cc50add00baddbdd59f378324da483/lib/plausible/google/http.ex#L128 

Your Plausible instance can't seem to resolve DNS records for `www.googleapis.com`
I don't know enough about Nix to help any further, sorry.

**Reply by anpin** [reply]
thanks @ruslandoga you were correct: I had some NAT issues preventing my instance to access internet

**Comment by raabmar** [comment]
How can we push this to get clearer instructions? I am searching for two weeks to find a way to enable Google Search Console, obviously some managed to get it to work somehow. How?

**Reply by wheeleran** [reply]
It does actually work, but the docs are misleading so use them as an indication only. @jonahgreenthal is right about domain verification, as far as I can tell, but my Search Console results are up and working for a few months now, without needing to reverify.

**Reply by ruslandoga** [reply]
I'll try adding Google Search Console integration to one of my websites tomorrow. I'll update the guide if it works out :)

**Reply by raabmar** [reply]
I think I spent already 10 to 15 hours on it in total, running in circles, and can't get the Search Console integration. Do you know where the screenshot of @jonahgreenthal comes from? I couldn't see those settings anywhere

1. Do you know if the web app must be in Testing or Published?
2. Do you know if user type is external or internal?
3. Who is the test user? I used just my main Google Account that also hosts the Google Search Console as a test user
4. In the oAuth Consent configuration which domains do I add as Authorized Domains? All the individual website URLs as well or only the one for self-hosted Plausible e.g. https://plausible.mydomain.com
5. What did you add for "Application home page"? I just added https://plausible.mydomain.com
6. What about the plausible-conf.env? I just added the client ID and client secret like that, it does not say anywhere how to add it
GOOGLE-CLIENT-ID = 123456789
GOOGLE-CLIENT-SECRET = 123456789
7. Is this the right way as it states in the documentation to force the new plausible-conf.env configuration:
`docker-compose down --remove-orphans && docker-compose up -d`

All those instructions are very very misleading and in the tickets people are also running in circles for months unfortunately. 
I am kind of waiting for the lucky punch to randomly enable some hidden setting to make it work.

**Reply by ruslandoga** [reply]
Here're the values for an app I setup in September last year

1. Testing
2. External

<img width="200" alt="Screenshot 2023-11-27 at 01 32 01" src="https://github.com/plausible/analytics/assets/67764432/bcce467e-4387-495b-b7c1-84bf821bc8fa">

4. I used my Google Account user
5. I added `copycat.fun` for my instance at `who.copycat.fun`, the form warns that the domain "must be a top private domain" if you type an invalid domain, `plausible.mydomain.com` is invalid

<img width="300" alt="Screenshot 2023-11-27 at 01 32 33" src="https://github.com/plausible/analytics/assets/67764432/d0f8208c-7f98-4c04-b624-83b879a7cba4">
<img width="300" alt="Screenshot 2023-11-27 at 01 32 38" src="https://github.com/plausible/analytics/assets/67764432/d46e25e8-0409-4f00-865e-6bd1f9bd7a37">

6. Nothing, in my case these fields are empty

<img width="300" alt="Screenshot 2023-11-27 at 01 32 47" src="https://github.com/plausible/analytics/assets/67764432/6f9a63f8-fdae-4b3b-b0b7-ef0c24ab9bbb">

7. Yes, just add these env vars to the envfile, note however that it's not dashes but underscores, https://plausible.io/docs/self-hosting-configuration#google-api-integration
8. If you didn't provide any extra compose files with `-f` initially, then yes, that command would work

**Reply by raabmar** [reply]
Success! In the end as you can see just from my points there are lots of lots of possible issues. One single wrong setting can mess up everything. I switched back again to External. At point 4 I did add just the main domain without subdomain (I think that was a mistake when I listed the 7 points, I had it already like that previously)

What solved it was the underscore instead of dashes - thanks for pointing that out. But it does not say it anywhere in the documentation to even add it this way, so it was all just a good guess from my end
`GOOGLE_CLIENT_ID = 123456789`
`GOOGLE_CLIENT_SECRET = 123456789`

![2023-11-26_003657@2x](https://github.com/plausible/analytics/assets/15885290/20350e7b-268f-42f6-a2e7-126f3b4e8286)

I appreciate your valuable input, perhaps it's really possible to update the documentation a bit. I think it's not just me, but many people over many months always struggling with the Google Search Console integration

**Reply by raabmar** [reply]
Btw do you think it was a bad idea to call it plausible.mydomain.com (because of possible adblockers?)

**Reply by 0PandaDEV** [reply]
I have the same issue but managed to set it up correctly according to the docs and your corrected instructions, but it still gives me a rediret_uri_mismatch when I try to log in. Also, what am I now supposed to input at authorized domains my instance is on plausible.pandadev.net so do I have to input pandadev.net or with the subdomain because when I do it with the subdomain it gives me an error so what am I doing wrong.
![image](https://github.com/plausible/analytics/assets/70103896/15199b58-825c-42eb-b5e9-28c945f2a21c)
![image](https://github.com/plausible/analytics/assets/70103896/b92f917d-aaf0-4b8d-894a-555172a65035)

**Reply by ruslandoga** [reply]
👋 @0PandaDEV 

In my case, for a Plausible instance at `who.copycat.fun`, I have the following:

<img width="500" alt="Screenshot 2023-11-27 at 12 47 49" src="https://github.com/plausible/analytics/assets/67764432/f4bda751-87b8-4199-ac75-2b0ead3fc004">

<img width="500" alt="Screenshot 2023-11-27 at 12 47 02" src="https://github.com/plausible/analytics/assets/67764432/fd6a6d4e-ff85-4ef8-a6c6-e78557ae0f18">

**Reply by ruslandoga** [reply]
I went through Search Console integration again and was able to set it up following the documentation.

I'm not sure what exactly is causing problems, I guess making the guide more linear would be helpful? Right now it's split between [two](https://plausible.io/docs/google-search-console-integration) [articles.](https://plausible.io/docs/self-hosting-configuration#google-api-integration) Otherwise, all the necessary info is there.

**Reply by 0PandaDEV** [reply]
I think I found the problem in the plausible conf file I had BASE_URL set to http://localhost:8000 now I changed it to https://plausible.pandadev.net, and now it works kind of because now it gives me an error after authenticating with it 
<img width="843" alt="image" src="https://github.com/plausible/analytics/assets/70103896/e15177ec-1c3b-4f02-ac45-59c099f09292">

**Reply by xeroc** [reply]
I am getting the same read warning. Nothing resolved this yet. Did you get this to work somehow?

**Reply by 0PandaDEV** [reply]
No unfortunately not

**Reply by wheeleran** [reply]
> I went through Search Console integration again and was able to set it up following the documentation.
> 
> I'm not sure what exactly is causing problems, I guess making the guide more linear would be helpful? Right now it's split between [two](https://plausible.io/docs/google-search-console-integration) [articles.](https://plausible.io/docs/self-hosting-configuration#google-api-integration) Otherwise, all the necessary info is there.

Part of it is the language used, for instance: the mentions of 'verify domain.' The instructions describe sidebar links and menu items that just aren't there. So if it's actually just adding the TLD and that's it, the instructions should say that. I got it working months ago, and this is an invaluable thread because you're working though the the setup, confirming what some of us have done successfully, and documenting it for those that got stuck!

**Reply by ruslandoga** [reply]
👋 @wheeleran 

We are re-doing the self-hosted docs and I wonder what you think about the Google guide at https://github.com/plausible/community-edition/wiki/google-integration

**Reply by 0PandaDEV** [reply]
Thank you so much, it helped me resolve this issue finally after about 4 months. It's much better when docs are with screenshots and step-by-step instructions. Btw I hate docs :)

**Reply by grssnbchr** [reply]
I just did this (did it over a year ago with the old instructions). I have to say it works like a charm, thanks for the screenshots!
Just one question: what ways are there to circumvent the 7 days restriction?
do I need to have to make the effort to tell Google why I want to publish my app?
![image](https://github.com/user-attachments/assets/ccdea412-1302-4450-afa2-5305e7b3d44e)

**Reply by ruslandoga** [reply]
👋 @grssnbchr 

I don't think I had any 7 days restrictions on my account. I still use an unpublished "external" version of the Google OAuth app.

**Reply by grssnbchr** [reply]
Weird, after 7 days, I get this:

![image](https://github.com/user-attachments/assets/6269ba92-cead-4c5b-a55f-83abb32b0e15)

When checking the [docs](https://plausible.io/docs/google-search-console-integration#i-get-the-invalid-grant-error), none of these points apply to me:

- Google has a limit of 50 live tokens. When you try to add the 51st site, the site you added first will be disconnected from your Plausible account and so on. This is a limitation on Google's side so if you have more than 50 different sites in your Search Console, you will need to split them between different Google accounts to get them into Plausible at the same time
- Your permissions for the specific site may have been changed or revoked in your Google account. Please check your Search Console account and make sure you still have the user permission to the site in question-
- Your token is inactive as your account has expired or your site isn't active. Please check your Search Console account and make sure you still have the account and are still tracking the site in question correctly
- You may have changed the permissions or the password in your Google account since you gave us the access. Please make sure to reconnect your Google account again to Plausible with the correct account details


What else could it be?

The workaround is to unlink and re-link my Google Account

**Comment by ruslandoga** [comment]
Domain verification happens through [Google Search console,](https://search.google.com/u/1/search-console) documented [here.](https://plausible.io/docs/google-search-console-integration)

Domain verification is required :)

**Reply by wheeleran** [reply]
Could I suggest that "domain verification required" is what created all the confusion in the first place, and the instructions for it were unclear at best, misleading even, because what the docs describe just isn't there. Please be literal in the instructions, to add client_ID and client_secret to docker-compose, and to specify the TLD example.com as the verifiable domain even if plausible is on a subdomain. There's no verifying to be done, it just has to be set up properly.

**Reply by ruslandoga** [reply]
Thank you for the suggestions! "Domain verification" under "APIs & Services" is indeed no longer accessible. I'll change that paragraph to point to Google Search console domain verification instead. The other two points are already present in the documentation. If the wording is not literal enough, please consider opening a PR with something better :)

**Reply by ruslandoga** [reply]
> There's no verifying to be done, it just has to be set up properly.

It's not required if Google Search property is already set up, otherwise the new users would need to add a domain there and verify it.
